### PR TITLE
Update and streamline method and type names

### LINKED ISSUE #580: Namings
This a list of type names, which I think could be improved upon:

* DataService: `isRelevantKey` -> `needsNotificationAbout`? 
* DataService: `initialRootRun` -> `initDataChangeListeners`
* DataService: `notifyChange` -> `notifyDataChangeListeners`
* DataServiceListener: `notifyChange` (same name as DataService) -> `dataChanged`?
* RuntimeListener: `notify` -> `runtimeChanged` 
* BaseControl: `inputElement` -> `createInputElement` as the contract seems to be to return a new element on each invocation, do not model as property
* Rename `JsonFormsHolder` to `JsonForms`
* Rename `JsonForms` to `JsonFormsElement` (as the actual element name does not matter much anyways)
* `JsonForms` has almost same methods as a `Renderer`. Try to unify both types, if not possible rename methods in `JsonForms` to resemble those of `Renderer` 

Consistency-related issues

* Capital usage: `UISchemaElement` vs `UiSchemaTester`/`UiSchemaRegistry` etc. 
* When finding elements via services/registries: most fitting vs. best vs. applicable
  * `UISchemaRegistry`: `getBestUiSchema` -> `findMostApplicabeUISchema`
* Consistent usage of `register`/`deregister` -> follow-up issue #583

### PR DESCRIPTION
**Pull request by edgarmueller**
Fixes #580:
* Also updates test descriptions

### CODE REVIEW

File: src/core.ts [hunk 1]
```diff
@@ -36,15 +36,15 @@ export interface JsonFormsServiceConstructable {
...
 export const JsonFormsServiceElement = (config) => (cls: JsonFormsServiceConstructable) => {
-  JsonFormsHolder.jsonFormsServices.push(cls);
+  JsonForms.jsonFormsServices.push(cls);
 };
 
 /**
  * Global JSONForms object that holds services and registries.
  */
-export class JsonFormsHolder {
+export class JsonForms {
```
Comment (eneufeld) [inline src/core.ts]: JsonForms is also not a perfect name. Is this maybe a JsonFormsConfiguration?
But as I don't have a better name for this, I'm fine with it, we just should think about how to name the global variable that is created by webpack

File: src/core/data.service.ts [hunk 2]
```diff
@@ -43,18 +44,18 @@ export class DataService {
...
    * Notifies any data change listeners about a data change.
    *
-   * @param {ControlElement} uischema the affected control element
+   * @param {ControlElement} controlElement the affected control element
    * @param newValue the new value of the data chunk
    */
-  notifyChange(uischema: ControlElement, newValue: any): void {
-    if (uischema !== undefined && uischema !== null) {
-      const pair = getValuePropertyPair(this.data, uischema.scope.$ref);
+  notifyDataChangeListeners(controlElement: ControlElement, newValue: any): void {
```
Comment (eneufeld) [inline src/core/data.service.ts]: In my opinion this name is not correct, as we are calling this on the dataservice in order to notify the dataservice about a datachange, so if a control renderer changes a value it notifies the dataservice.
So a better name would be notifyAboutDataChange ?

File: src/core/data.service.ts [hunk 3]
```diff
@@ -8,29 +8,30 @@ export interface DataChangeListener {
...
    * Represents the listener's logic to be called in case of a data change that
    * is relevant to the listener.
    *
-   * @param {ControlElement} uischema the control element that is affected by the data change
+   * @param {ControlElement} uiSchema the control element that is affected by the data change
    * @param {any} newValue the changed data value
    * @param {any} data the current data value
    */
-  notifyChange(uischema: ControlElement, newValue: any, data: any): void;
+  dataChanged(uiSchema: ControlElement, newValue: any, data: any): void;
```
Comment (eneufeld) [inline src/core/data.service.ts]: should we name "uiSchema" -> "controlElement"?

File: src/core/data.service.ts [hunk 4]
```diff
@@ -63,26 +64,26 @@ export class DataService {
...
-    this.changeListeners.splice(this.changeListeners.indexOf(listener), 1);
+  deregisterDataChangeListener(listener: DataChangeListener): void {
+    this.dataChangeListeners.splice(this.dataChangeListeners.indexOf(listener), 1);
   }
 
   /**
    * Resolve the ref of the given control against the root data.
    *
-   * @param {ControlElement} uischema
+   * @param {ControlElement} uiSchema
```
Comment (eneufeld) [inline src/core/data.service.ts]: uiSchema -> controlElement

File: src/core/data.service.ts [hunk 5]
```diff
@@ -63,26 +64,26 @@ export class DataService {
...
    * @return {any} the de-referenced data chunk
    */
-  getValue(uischema: ControlElement): any {
-    const pair = getValuePropertyPair(this.data, uischema.scope.$ref);
+  getValue(uiSchema: ControlElement): any {
```
Comment (eneufeld) [inline src/core/data.service.ts]: uiSchema -> controlElement

File: src/core/uischema.registry.ts [hunk 6]
```diff
@@ -54,44 +54,44 @@ export interface UiSchemaTester {
...
  */
-const NO_UISCHEMA_DEFINITION = {uiSchema: null, tester: (schema) => 0} as UiSchemaDefinition;
+const NO_UISCHEMA_DEFINITION = {uiSchema: null, tester: (schema) => 0} as UISchemaDefinition;
 
 /**
  * Implementation of the UI schema registry.
  */
-export class UiSchemaRegistryImpl implements UiSchemaRegistry {
-    private registry: Array<UiSchemaDefinition> = [];
+export class UiSchemaRegistryImpl implements UISchemaRegistry {
```
Comment (eneufeld) [inline src/core/uischema.registry.ts]: UiSchemaRegistryImpl -> UISchemaRegistryImpl

File: src/renderers/additional/array-renderer.ts [hunk 7]
```diff
@@ -41,15 +41,15 @@ export class ArrayControlRenderer extends Renderer implements DataChangeListener
   /**
    * @inheritDoc
    */
-  isRelevantKey (uischema: ControlElement): boolean {
+  needsNotificationAbout (uischema: ControlElement): boolean {
```
Comment (eneufeld) [inline src/renderers/additional/array-renderer.ts]: uiSchema -> controlElement

File: src/renderers/controls/base.control.ts [hunk 8]
```diff
@@ -80,21 +80,21 @@ export abstract class BaseControl <T extends HTMLElement>
...
-    this.dataService.unregisterChangeListener(this);
+    this.dataService.deregisterDataChangeListener(this);
     super.disconnectedCallback();
   }
 
   /**
    * @inheritDoc
    */
-  isRelevantKey (uischema: ControlElement): boolean {
+  needsNotificationAbout (uischema: ControlElement): boolean {
```
Comment (eneufeld) [inline src/renderers/controls/base.control.ts]: uiSchema -> controlElement

File: src/renderers/controls/base.control.ts [hunk 9]
```diff
@@ -104,7 +104,7 @@ export abstract class BaseControl <T extends HTMLElement>
   /**
    * @inheritDoc
    */
-  notifyChange(uischema: ControlElement, newValue: any, data: any): void {
+  dataChanged(uischema: ControlElement, newValue: any, data: any): void {
```
Comment (eneufeld) [inline src/renderers/controls/base.control.ts]: uiSchema -> controlElement

File: src/renderers/controls/base.control.ts [hunk 10]
```diff
@@ -161,7 +161,7 @@ export abstract class BaseControl <T extends HTMLElement>
    *
    * @returns {T} the created HTML input element
    */
-  protected abstract get inputElement(): T;
+  protected abstract get createInputElement(): T;
```
Comment (eneufeld) [inline src/renderers/controls/base.control.ts]: should not be a getter maybe?

File: test/json-forms.test.ts [hunk 11]
```diff
@@ -23,17 +22,17 @@ test('jsonforms only data', t => {
...
+test('Connect JSON Forms element with data and data schema set', t => {
+    const jsonForms = new JsonFormsElement();
     jsonForms.data = {name: 'foo'};
     const dataSchema = {type: 'object', properties: {name: {type: 'string'}}} as JsonSchema;
     jsonForms.dataSchema = dataSchema;
     jsonForms.connectedCallback();
     t.is(jsonForms.children.length, 0);
 });
-test.cb('jsonforms data and schema', t => {
+test.cb('Connect JSON Forms element asynchronously with data and data schema set', t => {
```
Comment (eneufeld) [inline test/json-forms.test.ts]: the async comes from the dataschema resolution, so it is just needed to run the test successfully

File: src/core/data.service.ts (hunk 2 above)
Comment (edgarmueller) [inline src/core/data.service.ts]: Hm, yes, true, I focused too much on the listener aspect here. Actually, we should probably rename and split this up to:

```JS
changeData(ControlElement, any) {
  // assign newValue as before
  notifyDataChangeListeners()
}
```

What do you think?

File: src/core/data.service.ts (hunk 3 above)
Comment (edgarmueller) [inline src/core/data.service.ts]: Yes, I missed this one.

File: test/json-forms.test.ts (hunk 11 above)
Comment (edgarmueller) [inline test/json-forms.test.ts]: I know, but I couldn't think of any other name. Suggestions?

File: src/core/data.service.ts (hunk 2 above)
Comment (eneufeld) [inline src/core/data.service.ts]: I agree, the notifyDataChangeListeners() would be private then?

File: test/json-forms.test.ts (hunk 11 above)
Comment (eneufeld) [inline test/json-forms.test.ts]: can''t we simply omit the async text in the test? or we make all of them async

File: src/core/data.service.ts [hunk 12]
```diff
@@ -8,29 +8,30 @@ export interface DataChangeListener {
...
   /**
    * Determines whether this listener is interested in any data change.
-   * If this returns true, the notifyChange method of the listener will be called.
-   * @param {ControlElement} uischema the control element that is affected by the data change
+   * If this returns true, the dataChanged method of the listener will be called.
+   * @param {ControlElement} uiSchema the control element that is affected by the data change
    * @returns whether this listener is interested in the given control element
    */
-  isRelevantKey(uischema: ControlElement): boolean;
+  needsNotificationAbout(uiSchema: ControlElement): boolean;
```
Comment (eneufeld) [inline src/core/data.service.ts]: uiSchema -> controlElement

File: src/renderers/additional/array-renderer.ts [hunk 13]
```diff
@@ -41,15 +41,15 @@ export class ArrayControlRenderer extends Renderer implements DataChangeListener
...
+  needsNotificationAbout (controlElement: ControlElement): boolean {
+    return controlElement === undefined || controlElement === null
+    ? false : (<ControlElement>this.uischema).scope.$ref === controlElement.scope.$ref;
   }
 
   /**
    * @inheritDoc
    */
-  notifyChange(uischema: ControlElement, newValue: any, data: any): void {
+  dataChanged(uischema: ControlElement, newValue: any, data: any): void {
```
Comment (eneufeld) [inline src/renderers/additional/array-renderer.ts]: uischema -> controlElement

File: src/renderers/additional/table-array.control.ts [hunk 14]
```diff
@@ -42,15 +42,15 @@ export class TableArrayControlRenderer extends Renderer implements DataChangeLis
   /**
    * @inheritDoc
    */
-  isRelevantKey (uischema: ControlElement): boolean {
+  needsNotificationAbout (uischema: ControlElement): boolean {
```
Comment (eneufeld) [inline src/renderers/additional/table-array.control.ts]: uischema -> controlElement

File: src/renderers/additional/table-array.control.ts [hunk 15]
```diff
@@ -42,15 +42,15 @@ export class TableArrayControlRenderer extends Renderer implements DataChangeLis
...
     return uischema === undefined || uischema === null
     ? false : (<ControlElement>this.uischema).scope.$ref === uischema.scope.$ref;
   }
 
   /**
    * @inheritDoc
    */
-  notifyChange(uischema: ControlElement, newValue: any, data: any): void {
+  dataChanged(uischema: ControlElement, newValue: any, data: any): void {
```
Comment (eneufeld) [inline src/renderers/additional/table-array.control.ts]: uischema -> controlElement

File: src/renderers/additional/tree-renderer.ts [hunk 16]
```diff
@@ -126,12 +126,12 @@ export class TreeMasterDetailRenderer extends Renderer implements DataChangeList
     this.renderFull();
     return this;
   }
-  isRelevantKey (uischema: ControlElement): boolean {
+  needsNotificationAbout (uischema: ControlElement): boolean {
```
Comment (eneufeld) [inline src/renderers/additional/tree-renderer.ts]: uischema -> controlElement

File: src/renderers/additional/tree-renderer.ts [hunk 17]
```diff
@@ -126,12 +126,12 @@ export class TreeMasterDetailRenderer extends Renderer implements DataChangeList
...
     return uischema === undefined || uischema === null ? false :
       (<ControlElement>this.uischema).scope.$ref === uischema.scope.$ref && !this.addingToRoot;
   }
 
-  notifyChange(uischema: ControlElement, newValue: any, data: any): void {
+  dataChanged(uischema: ControlElement, newValue: any, data: any): void {
```
Comment (eneufeld) [inline src/renderers/additional/tree-renderer.ts]: uischema -> controlElement

File: src/renderers/additional/tree-renderer.ts [hunk 18]
```diff
@@ -311,16 +311,16 @@ export class TreeMasterDetailRenderer extends Renderer implements DataChangeList
...
 
-    const jsonForms = <JsonForms>document.createElement('json-forms');
+    const jsonForms = <JsonFormsElement>document.createElement('json-forms');
     jsonForms.data = element;
     jsonForms.dataSchema = schema;
     // check needed for tests
     if (jsonForms.addDataChangeListener) {
       jsonForms.addDataChangeListener({
-        isRelevantKey: (uischema: ControlElement): boolean => {
+        needsNotificationAbout: (uischema: ControlElement): boolean => {
```
Comment (eneufeld) [inline src/renderers/additional/tree-renderer.ts]: uischema -> controlElement

File: src/renderers/additional/tree-renderer.ts [hunk 19]
```diff
@@ -311,16 +311,16 @@ export class TreeMasterDetailRenderer extends Renderer implements DataChangeList
...
           return uischema !== null;
         },
-        notifyChange: (uischema: ControlElement, newValue: any, data: any): void => {
+        dataChanged: (uischema: ControlElement, newValue: any, data: any): void => {
```
Comment (eneufeld) [inline src/renderers/additional/tree-renderer.ts]: uischema -> controlElement

**Pull request merged**